File: first-unique-character-in-a-string/shaun-hash.cpp
//shaun own idea and code, accepted after 1 typo fix
//runtime 32ms, faster than 98%, mem less than 88%
class Solution {
public:
    int firstUniqChar(string s) {
        //Note: You may assume the string contain only lowercase letters.
        //vector<int> m(26);
        int m[26] = {0};
        for (char ch: s) m[ ch - 'a']++;
        
        for (int i = 0; i < s.length(); ++i) {
            if (m[s[i] - 'a'] == 1) return i;
        }
        return -1;
    }
};

File: first-unique-character-in-a-string/shaun-hash.java
/*

Runtime: 11 ms, faster than 60.99% of Java online submissions for First Unique Character in a String.
Memory Usage: 47.3 MB, less than 5.71% of Java online submissions for First Unique Character in a String.

*/
class Solution {
    public int firstUniqChar(String s) {
        int [] m = new int[26]; //lower case letters
        char [] chars = s.toCharArray();
        for (int i = 0; i < chars.length; i++) {
            char c = chars[i];
            m[c - 'a'] += 1;
        }

        for (int i = 0; i < chars.length; i++) {
            char c = chars[i];
            if (m[c - 'a'] == 1) return i;
        }
        return -1;

    }
}
